### DISCUSSION TITLE
how to expand `uses` statements in pyang-style tree

### DISCUSSION THREAD

**Discussion by hellt** [CESNET/libyang#1866] [Q&A] [answered]
I am trying to produce a pyang-like tree using yanglint. Unfortunately, the produced tree renders to be quite terse, as it displays the grouping name, and doesn't expand this group any further

```
❯ docker run --rm -it -v $(pwd):/work ghcr.io/hellt/yanglint -p yang/modules/models/srl_nokia/models yang/modules/models/srl_nokia/models/system/srl_nokia-event-handler.yang  -f tree

libyang warn: Must condition "/srl_nokia-if:interface[srl_nokia-if:name=substring-before(current(), '.')]/srl_nokia-if:subinterface[srl_nokia-if:index=substring-after(current(), '.')]/srl_nokia-if:type != 'local-mirror-dest'" check skipped because referenced module "srl_nokia-interfaces" is not implemented.
libyang warn: Schema node "proxy-arp" for parent "/srl_nokia-network-instance:network-instance/bridge-table" not found; in expr "not(starts-with(.,'irb') and ../../bridge-table/proxy-arp" with context node "/srl_nokia-network-instance:network-instance/interface/name".

module: srl_nokia-event-handler

  augment /srl-system:system:
    +---u eh-top
```

**Comment by michalvasko** [comment, marked as answer]
All the module in the `tree` output need to be implemented, so in your case you probably want to add the target module of the augment to the command line. It will then print its `tree` output including the appended augment nodes.

**Reply by hellt** [reply]
I was first trying to force-implementing all modules by using `-ii` flag, which still produced a terse-ified output

```
yanglint -p yang/modules/models/srl_nokia/models yang/modules/models/srl_nokia/models/system/srl_nokia-event-handler.yang -ii -f tree
```

but then I think what you meant was to specify both the module in question and the one that it augments:

```
yanglint -p yang/modules/models/srl_nokia/models yang/modules/models/srl_nokia/models/system/srl_nokia-system.yang yang/modules/models/srl_nokia/models/system/srl_nokia-event-handler.yang -ii -f tree
```

The latter provides an expanded tree on a /system container, which includes also the `srl_nokia-event-handler` module.

So it seems yanglint is not capable of expanding the module as pyang does?

```
pyang -p yang/modules/models/srl_nokia --strict -f tree yang/modules/models/srl_nokia/models/system/srl_nokia-event-handler.yang      
module: srl_nokia-event-handler

  augment /srl-system:system:
    +--rw event-handler! {srl-feat:event-handler}?
       +--rw run-as-user?   -> /srl-system:system/srl-aaa:aaa/authentication/user/username
       +--rw instance* [name]
          +--rw name                      srl-comm:name
          +--rw admin-state?              srl-comm:admin-state
          +--rw trigger-on?               enumeration
          +--rw cron?                     string
          +--rw updates-only?             boolean
          +--rw network-instance?         -> /srl-netinst:network-instance/name
          +--rw timeout?                  uint64
          +--rw paths*                    string
          +--rw options
          |  +--rw object* [name]
--snip--
```

**Reply by michalvasko** [reply]
No, `yanglint` cannot print non-implemented nodes as `tree`, which nodes in a non-connected augment are.

**Reply by hellt** [reply]
thanks @michalvasko

**Reply by jktjkt** [reply]
Also, the order of modules listed on the command line matters. If I have a module `a.yang` which provides a container, then I augment it via `b.yang`, and finally augment the result via `c.yang`, then it makes sense to just run `yanglint -f tree a.yang b.yang c.yang`. If I only run it via `yanglint -f tree c.yang`, then only these raw augments will be printed.
